### ISSUE TITLE
bug(mac): Caps Lock rules are not behaving consistently between macOS and Windows

### ISSUE THREAD

**Issue by caforbes** [keymanapp/keyman#7632]
### Is your feature request related to a problem? Please describe.

I'm currently reworking the SENĆOŦEN desktop keyboard on request from the community. The new design is for the keyboard to be essentially bilingual, optimized to type in SENĆOŦEN with Caps Lock on and in English with Caps Lock off.

Windows and Mac seem to have very different default Capslock rules, as I've now discovered. I can produce the desired keyboard 100% on Windows with a pretty basic rule set. But on Mac, turning Capslock on breaks almost every rule.

`'T' + [K_SLASH] > 'Ⱦ'` -- with capslock on produces `Ⱦ` on Windows, but produces `T/` on Mac.

In order to reproduce the same effect on Mac, I am forced to duplicate all my rules in the [NCAPS] and [CAPS] states.

`'T' + [NCAPS K_SLASH] > 'Ⱦ'`
`'T' + [CAPS K_SLASH] > 'Ⱦ'` -- produces `Ⱦ` for both systems.

This is annoying and leads to a lot of warnings saying that I need to specify CAPS or NCAPS everywhere that various keys are used, just because I am trying to tell them that they should behave the same both in and out of capslock.

### Describe the solution you'd like

The `&CasedKeys` store is used to generate CAPS and NCAPS rules using "normal casing rules", essentially so that rules are inverted with their SHIFTed equivalent while Caps lock is on. [Existing docs](https://help.keyman.com/developer/language/reference/casedkeys)

I would love it if there were an equivalent `&UncasedKeys` store to designate keys with rules whose shifted or unshifted behavior should persist as-is into the caps lock state. This would be the "normal" caps lock behavior for most non-cased keys like punctuation or diacritics.

Desired end result would be to write the following, without having to specify duplicate CAPS and NCAPS rules:

`store(&UncasedKeys) [K_SLASH]`
`'T' + [K_SLASH] > 'Ⱦ'`

This would overall be a way to make keyboard behavior more consistent across platforms when there is an expectation that users will use capslock.

### Describe alternatives you've considered

I still don't completely understand the difference between Mac and Windows default capslock behavior and why exactly Macs treat capslock so differently, so it is difficult to speak to any more complex solution. Maybe there could be a developer setting in the layout for "use windows-like capslock" or "use mac-like capslock", which would let the compiler explicitly generate rules instead of relying on the user's default settings. This doesn't seem as transparent though.

### Keyman apps

- [ ] Keyman for Android
- [ ] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [X] Keyman for macOS
- [X] Keyman for Windows
- [X] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

15.0.269

### Operating system

macOS Monterey

### Keyboard name

fv_sencoten

### Keyboard version

9.1.1 - 9.2

### Language name

SENĆOŦEN

**Comment by mcdurdin** [comment]
The behaviour of caps lock *should* be the same on Windows and macOS, so if it isn't, there is definitely something there we need to follow up. Note that we are planning to replace the core of Keyman for macOS (#5012) in Keyman 17.0, and that should resolve most compatibility issues.

It would be helpful to produce a minimal keyboard example that illustrates the difference between Windows and macOS keyboard behaviour, and attach it to this issue.

**Comment by mcdurdin** [comment]
I am going to reclassify this as a potential Keyman for macOS bug; the `&uncasedkeys` store is implicitly the inverse of the `&casedkeys` store.

The warning that you receive is due to the fact that there is ambiguity about how the compiler groups rules to be applied for a given key if some rules have NCAPS/CAPS and others don't.

**Comment by caforbes** [comment]
Awesome, thanks for bring it to the real problem. As I kept hammering at this it became clear there is no way for me to do what I'm trying to do on Mac. I've tried it multiple ways and it's always fine on Windows and doesn't work at all on Mac.

Here's a sample minimal keyboard: [mwe_capslockbug.zip](https://github.com/keymanapp/keyman/files/9955728/mwe_capslockbug.zip)

```
# Windows works as desired
Windows basic       a   a'  a/  c   c'  c/
Windows shift       A   Á   Ⱥ   C   Ć   Ȼ
Windows caps        A   Á   Ⱥ   C   Ć   Ȼ
Windows caps+shift  a   a'  a/  c   c'  c/

# Mac is complete garbage, I can't get anything to show up
Mac basic       a   a'  a/  c   c'  c/
Mac shift       A   A   A   C   C   C
Mac caps        A   A'  A   C   C'  C
Mac caps+shift  a   a'  a/  C   C'  C
```

(for the shift states, I use Shift-Letter but regular Quote)

I've restarted my computer multiple times, uninstalled and reinstalled the kmp locally, uninstalled and reinstalled Keyman, and seems like it's just not producing anything on Mac once you start specifying CAPS and NCAPS. Before I began specifying CAPS and NCAPS, I could at least get it to work on the shift layer. Now I don't know what's going on.

The community has specifically requested a keyboard which works best with Capslock on, which at this point seems impossible (for Mac). Please keep me updated if there's any progress or if you have any suggestions.

**Comment by mcdurdin** [comment]
Thank you for the keyboard. I am not able to reproduce the problem here -- the rules are all working fine. This suggests that it might be that Keyman is having keystrokes blocked due to security settings on your mac. Can you follow the steps at https://help.keyman.com/products/mac/current-version/troubleshooting/configure-security to verify that there is nothing going wrong there?

A secondary thing to check is at https://help.keyman.com/knowledge-base/104

Note that Keyman 16 beta includes some additional security configuration verification which streamlines this process considerably. The security model changes with each release of macOS, so we have to keep adapting to it!

**Comment by caforbes** [comment]
Thanks so much, looks like the Accessibility one wasn't checked. Now it works fine. Glad it was such an easy fix -- sad that I wasted so much of my life on it. D:

**Comment by mcdurdin** [comment]
Hurrah! I'm glad that fixed the issue. I am sorry this issue wasted so much of your time. As I noted, the Keyman 16 release should make this work much better -- @sgschantz invested significant energy into validating the various scenarios in which Keyman may not receive correct permissions, across a variety of macOS releases, and building in automatic testing for the security status of Keyman.

**Comment by caforbes** [comment]
The original issue I raised still stands. Is there a simpler way to create consistent capslock behavior across Windows and Mac without writing every rule twice with CAPS and NCAPS? To reiterate:

`'T' + [K_SLASH] > 'Ⱦ'` -- with capslock on produces Ⱦ on Windows, but produces T/ on Mac.

`'T' + [NCAPS K_SLASH] > 'Ⱦ'`
`'T' + [CAPS K_SLASH] > 'Ⱦ'` -- produces Ⱦ for both systems.

**Comment by mcdurdin** [comment]
@caforbes sorry for the delayed response; I have been away for the last couple of days.

Do you have the following rules in your keyboard?

```
store(&UncasedKeys) [K_SLASH]
'T' + [K_SLASH] > 'Ⱦ'
```

**Comment by mcdurdin** [comment]
@caforbes just waiting for your reply on this so we can action it :grin:

**Comment by mcdurdin** [comment]
I'm really sorry, I did get this wrong. `&UncasedKeys` doesn't exist... because any key that isn't in `&CasedKeys` should be uncased.

This rule should match the slash key whether or not Caps Lock is on, on any platform:
```
'T' + [K_SLASH] > 'Ⱦ'
```

If it isn't matching, then, yes, we will need to chase that on macOS (I will try and test this soon here; are you able to attach your whole keyboard source folder as a zip so I can be sure I am testing the right thing?).

However, the fix on macOS will most likely be with the implementation of #5012, and the workaround will be to add the two rules for now, which you could do with a store to make it a little less messy:

```
store(SLASH) [NCAPS K_SLASH] [CAPS K_SLASH]
'T' + any(SLASH) > 'Ⱦ'
```

**Comment by mcdurdin** [comment]
Another repro test for my use:
[test_caps_ncaps_macos.zip](https://github.com/keymanapp/keyman/files/10071043/test_caps_ncaps_macos.zip)

**Comment by mcdurdin** [comment]
I've reproduced the issue here, and it is a bug in the legacy Keyman for macOS modifier matching. I'm working on a fix which should make it into 16.0-beta, because it looks like it is pretty straightforward and should be reliable. (We'll throw the fix away in 17.0 when we move to new Core, but that's okay!)
